torch/distributed/optim/named_optimizer.py

```diff
@@ -260,11 +260,28 @@ def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
 
         self._optimizer.load_state_dict(new_state_dict)
 
-    # pyre-ignore [2]
-    def add_param_group(self, param_group: Any) -> None:
-        raise NotImplementedError(
-            "add_param_group not supported yet and might be implemented soon."
+    def add_param_group(self, param_group: Mapping[str, Any]) -> None:
+        warnings.warn(
+            "This API is still in develop and it might not work as expected "
+            "under the distributed training scenario."
         )
+        assert isinstance(param_group, dict), "param group must be a dict"
+
+        params = param_group['params']
+        if isinstance(params, torch.Tensor):
+            param_group['params'] = [params]
+        else:
+            param_group['params'] = list(params)
+
+        param_to_key = {param: key for key, param in self.named_parameters.items()}  # type: ignore[misc, has-type]
+        for param in param_group['params']:
+            if param not in param_to_key:
+                raise ValueError("some parameters are not in the module")
```

Comment:
If this error is being hit, might be a more detailed error msg if we log all the parameter names that are not in the module.

Also, could we add a unittest that exercises this error path?

Reply:
Sure.

Reply:
After a second thoughts, this is actually tricky. Because the params added does not have name that's why the error msg are quite generic.